src/lib/staking/safety-module-helper.ts

```diff
@@ -0,0 +1,202 @@
+import * as ERC20ABI from '../abi/erc20.abi.json';
+import * as BVaultABI from '../abi/balancer-vault.abi.json';
+import * as MultiCallerABI from '../abi/multicaller.abi.json';
+import {
+  BalancerVaultAddress,
+  Balancer_80PSP_20WETH_address,
+  Balancer_80PSP_20WETH_poolId,
+  CHAIN_ID_MAINNET,
+  MulticallEncodedData,
+  MULTICALL_ADDRESS,
+  PSP_ADDRESS,
+  SAFETY_MODULE_ADDRESS,
+} from '../constants';
+import { Provider } from '../provider';
+import { Contract, BigNumber as EthersBN } from 'ethers';
+import { Interface } from '@ethersproject/abi';
+import { getTokenHolders } from '../utils/covalent';
+import { DataByAccount, PSPStakesForStaker, StkPSPBPtState } from './types';
+
+export class SafetyModuleHelper {
+  private static instance: SafetyModuleHelper;
+
+  static getInstance() {
+    if (!SafetyModuleHelper.instance) {
+      SafetyModuleHelper.instance = new SafetyModuleHelper();
+    }
+    return SafetyModuleHelper.instance;
+  }
+
+  chainId = CHAIN_ID_MAINNET; // all staking programs are only available on ethereum mainnet
+  multicallContract: Contract;
+  safetyModuleAsERC20: Contract;
+  bVaultIface: Interface;
+  erc20Iface: Interface;
+
+  constructor() {
+    const provider = Provider.getJsonRpcProvider(this.chainId);
+
+    this.multicallContract = new Contract(
+      MULTICALL_ADDRESS[this.chainId],
+      MultiCallerABI,
+      provider,
+    );
+
+    this.safetyModuleAsERC20 = new Contract(
+      SAFETY_MODULE_ADDRESS,
+      ERC20ABI,
+      provider,
+    );
+
+    this.bVaultIface = new Interface(BVaultABI);
+    this.erc20Iface = new Interface(ERC20ABI);
+  }
+
+  // this computes the staked PSP in the safety module for one account. This method is safe regarding to slashing.
+  getPSPStakedInSafetyModule = async (
+    account: string,
+  ): Promise<PSPStakesForStaker<string>> => {
+    const [
+      {
+        bptBalanceOfStkPSPBpt,
+        pspBalance,
+        stkPSPBPtTotalSupply,
+        bptTotalSupply,
+      },
+      stkPSPBalanceBN,
+    ] = await Promise.all([
+      this.fetchStkPSPBPtState(),
+      this.safetyModuleAsERC20.balanceOf(account) as Promise<EthersBN>,
+    ]);
+
+    const stkPSPBalance = stkPSPBalanceBN.toBigInt();
+
+    const totalPSPStaked =
+      (stkPSPBalance * bptBalanceOfStkPSPBpt * pspBalance) /
+      (stkPSPBPtTotalSupply * bptTotalSupply);
+
+    return {
+      pspStaked: totalPSPStaked.toString(),
+      breakdownByStakingContract: {
+        ...(!!totalPSPStaked && {
+          [SAFETY_MODULE_ADDRESS]: totalPSPStaked.toString(),
+        }),
+      },
+    };
+  };
+
+  /// functions to fetch multiple stakers data efficiently
+  async fetchStkPSPBptStakers(
+    blockNumber?: number,
+  ): Promise<DataByAccount<bigint>> {
```

Comment:
I must admit I've been a little juggling between bigint and bigNumber.
In a near future I'd like to move the whole code GRP dependant code to bigint but I'm a little bit afraid of the precision loss so I prefer not to for now and keep 🤹